Feat nf test guide 3 4 (#3586)
nf-core/website: sites/docs/src/content/docs/tutorials/tests_and_test_data/components/03_testing_modules.md

## Your first simple test walkthrough with an example

Let's walk through creating a basic test step-by-step using the `cat` module as an example:

### Step 1: Examine the module

First, let's look at a simple module (`modules/cat/main.nf`):

```groovy
process CAT {
    input:
    tuple val(meta), path(file_in)

    output:
    tuple val(meta), path("${prefix}"), emit: file_out
    path "versions.yml"               , emit: versions

    script:
    prefix = "${meta.id}.txt"
    """
    cat ${file_in} > ${prefix}
    """
}
```

This process simply concatenates input files and outputs a single file.

### Step 2: Write the test

Create `modules/cat/tests/main.nf.test`:

```groovy
nextflow_process {
    name "Test Process CAT"
    script "../main.nf"
    process "CAT"

    test("simple file concatenation") {
        when {
            process {
                """
                input[0] = [
                    [ id:'test' ], // meta map with sample identifier
                    file('test_file.txt')  // input file
                ]
                """
            }
        }

        then {
            assert process.success               // Check process completed successfully

>             assert snapshot(process.out).match() // Validate outputs match expected snapshot

Reviewer: ```suggestion
            assert snapshot(sanitizeOutput(process.out)).match() // Validate outputs match expected snapshot
```
Might be worth adding `sanitizeOutput` here in the docs too as it will create easier to read snapshots

        }
    }
}
```

**Understanding the test structure:**

The nf-test file is organized into several key scopes that define what and how to test:

- `nextflow_process { }` - **Test wrapper**: Declares this is a process-level test
- `name` - **Test description**: Human-readable name for the test suite

> - `script` - **Module location**: Absolute/Relative path to the `main.nf` file containing the process

Reviewer: ```suggestion
- `script` - **Module location**: Absolute/Relative path to the `main.nf` file containing the process. Relative paths should start from the location of the current test file.
```

- `process` - **Process name**: The specific process to test (must match the process name in `main.nf`)
- `test("...")` - **Individual test case**: Each test block defines one scenario to test
- `when { }` - **Test setup**: Where you define the input data for the process
- `then { }` - **Assertions**: Where you verify the process behaved as expected

**Inside the test components:**

Now that we understand the overall structure, let's look at the specific elements within the test:

- `[ id:'test' ]` - **meta map**: Contains sample metadata (ID, conditions, etc.)
- `file('test_file.txt')` - **input file**: The file to process
- `process.success` - **success check**: Ensures the process completed without errors
- `snapshot(process.out).match()` - **output validation**: Compares all outputs to stored snapshot

### Step 3: Run the test

Once you've written the test, you can generate the initial snapshot by running the test.

```bash
cd path/to/modules
nf-core modules test cat --profile docker (EXAMPLE)

> ```

Reviewer: Maybe not put `(EXAMPLE)` behind this as it looks like it's part of the command.

- `nextflow_process { }` - **Test wrapper**: Declares this is a process-level test
- `name` - **Test description**: Human-readable name for the test suite
- `script` - **Module location**: Absolute/Relative path to the `main.nf` file containing the process. Relative paths should start from the location of the current test file.
- `process` - **Process name**: The specific process to test (must match the process name in `main.nf`)
- `test("...")` - **Individual test case**: Each test block defines one scenario to test
- `when { }` - **Test setup**: Where you define the input data for the process
- `then { }` - **Assertions**: Where you verify the process behaved as expected

**Inside the test components:**

Now that we understand the overall structure, let's look at the specific elements within the test:

- `[ id:'test' ]` - **meta map**: Contains sample metadata (ID, conditions, etc.)
- `file('test_file.txt')` - **input file**: The file to process
- `process.success` - **success check**: Ensures the process completed without errors
- `snapshot(sanitizeOutput(process.out)).match()` - **output validation**: Compares all outputs to stored snapshot

> - `sanitizeOutput()` - **snapshot sanitization**: Cleans process and workflow outputs by removing numbered keys, making snapshots more human-readable

Reviewer 2: This seems overly specific and confusing (see above)
```suggestion
```

5. Reports PASS/FAIL

### Step 4: Examine the generated snapshot

After running, check `tests/main.nf.test.snap`:

```json
{
  "simple file concatenation": {
    "content": [
      {
        "0": [
          // First output channel (file_out)
          [{ "id": "test" }, "test.txt:md5,d41d8cd98f00b204e9800998ecf8427e"]
        ],
        "1": [
          // Second output channel (versions)
          "versions.yml:md5,c50aa59475ab20752531545a585f8f2d"
        ]
      }
    ]
  }
}
```

You can see the the outputs of the module, with an associated `md5sum` which concisely describes the exact contents of the file.

Now you understand the full test cycle! The snapshot ensures your module produces consistent, expected outputs every time.

Any time you change the module - you can run the test again to check to see you did not unexpectedly change an output that should not have been changed.

## Essential assertions

Tests use assertions to verify the expected output of the process specified in the `then` block.

You can specify multiple assertions to be evaluated together in a single test by specifying them within an `assertAll` block.

> Nextflow process output channels that lack explicit names (i.e., when no [`meta` map](https://nf-co.re/docs/contributing/components/meta_map) is present) can be addressed using square brackets and the corresponding index, for example `process.out[0]` for the first channel and `process.out[0][1]` for the second element of the first channel.

Reviewer 2: Uh this is confusing now, the `meta` map doesn't influence the explicit names. Maybe rephrase slightly... maybe more 'channels with multple objects in a tuple) or something.

assert process.out.my_channel != null
assert process.out.my_channel.size() == 3
assert process.out.my_channel.get(0) == "expected_value"

// For unnamed channels, use index notation
assert process.out[0] != null
assert process.out[0].size() == 3

// Group assertions to see all failures at once
assertAll(
    { assert process.success },
    { assert snapshot(process.out).match() }
)
```

:::note
For more nf-test assertion patterns, see the [nf-test assertions examples documentation](./07_assertions.md).
:::

## Creating a new module with tests

Now we can dive deeper into specifically making an nf-core module.

When creating a new module using `nf-core/tools`, a test file is automatically generated based on the template.

```bash
# Create a new module using nf-core/tools (EXAMPLE)
cd path/to/modules
nf-core modules create seqtk/sample
```

This creates the following module directory structure:

```tree
modules/nf-core/tool/subtool/
├── main.nf              # Process definition
├── meta.yml             # Module metadata
└── tests/               # Testing directory
    ├── main.nf.test     # Test definitions
    └── nextflow.config  # Optional: test-specific config
```

> We can then edit the test file (`tests/main.nf.test`) to specify the test's input data via the `input[0]` and `input[1]` channels in the `when` block.

Reviewer 2: ```suggestion
After writing the module under `main.nf` and being ready to write the tests, we can edit the test file (`tests/main.nf.test`) to specify the test's input data via the `input[0]` and `input[1]` channels in the `when` block.
```

Then import the config to the `main.nf.test` file and supply the params in the `when` section of the test.

```main.nf.test
process "MODULE"
config "./nextflow.config"

when {
  params {
    module_args = '--extra_opt1 --extra_opt2'
  }
  process {
    """
    input[0] = [
      [ id:'test1', single_end:false ], // meta map
      file(params.modules_testdata_base_path + 'genomics/prokaryotes/bacteroides_fragilis/genome/genome.fna.gz', checkIfExists: true)
    ]
    """
  }
}
```

## Testing chained modules

In some cases, rather than directly linking to pre-made test-data files within the `when:` scope, it may make sense to run an 'upstream module' in your test that produces the required inputs of the module you want to test.

The `setup` method allows you to specify processes or workflows that need to be executed before the primary `when` block.
It serves as a mechanism to prepare the required input data or set up essential steps prior to the primary processing block 'on the fly'.

Within a `setup` block, you can use the `run` method to define and execute multiple dependent processes or workflows.

> Here's a basic example of how a setup block looks:

Reviewer 2: ```suggestion
Here's a basic dummy example of how a setup block looks:
```
